api/src/wfl/module/covid.clj

```diff
@@ -23,6 +24,61 @@
 (defn create-covid-workload!
   [tx {:keys [source sink executor] :as request}])
 
+(defn start-covid-workload!
+  [{:keys [submission-id workspace] :as workload}])
+
+(defn ^:private snapshot-error-map
+  "Return default ExceptionInfo map for snapshot-related exceptions."
+  [{:keys [items sink] :as _workload} {:keys [id snapshot_id] :as _record}]
+  {:workspace (:workspace sink)
+   :table items
+   :id id
+   :snapshot_id snapshot_id})
+
+(defn ^:private try-get-snapshot-reference
+  "Attempt to fetch snapshot reference with id SNAPSHOT_REFERENCE_ID
+  from SINK-derived workspace."
+  [{:keys [sink] :as workload} {:keys [snapshot_reference_id] :as record}]
+  (let [workspace (:workspace sink)]
+    (try
+      (rawls/get-snapshot-reference workspace snapshot_reference_id)
+      (catch Throwable cause
+        (throw (ex-info "Rawls unable to fetch snapshot reference"
+                        (util/deep-merge
+                         (snapshot-error-map workload record)
+                         {:snapshot_reference_id snapshot_reference_id})
+                        cause))))))
+
+(defn ^:private snapshot-imported?
+  "Check if snapshot with id SNAPSHOT_ID in TDR has been imported to
+  SINK-derived workspace and can be successfully fetched."
+  [workload {:keys [snapshot_id snapshot_reference_id] :as record}]
+  {:pre [(some? snapshot_id)]}
+  (and (some? snapshot_reference_id)
+       (some? (try-get-snapshot-reference workload record))))
+
+(defn ^:private import-snapshot!
+  "Import snapshot with id SNAPSHOT_ID from TDR to workspace.
+  Use transaction TX to update ITEMS table with resulting reference id."
+  [tx {:keys [items sink] :as workload} {:keys [id snapshot_id] :as record}]
+  (let [workspace (:workspace sink)
+        name (util/randomize "placeholder-snapshot-ref-name")
+        reference-id (rawls/create-snapshot-reference workspace snapshot_id name)]
+    (when (nil? reference-id)
+      (throw (ex-info "Rawls unable to create snapshot reference"
+                      (util/deep-merge (snapshot-error-map workload record)
+                                       {:name name}))))
+    (jdbc/update! tx items {:snapshot_reference_id reference-id} ["id = ?" id])))
```

Comment:
Our schema def'n for `Sarscov2IlluminaFull` user-defined type has an `updated` column:
```
            updated                     timestamptz  -- status update time
```

Do we wish to update this value every time we modify row(s) in the table or only when workflow status is updated?

Reply:
```clojure
(util/deep-merge (snapshot-error-map workload record)
                 {:name name})
```
is probably just
```clojure
(assoc (snapshot-error-map workload record) :name name)
```

Reply:
Oh nice, I like this better here.

Reply:
> Our schema def'n for `Sarscov2IlluminaFull` user-defined type has an `updated` column:
> 
> ```
>             updated                     timestamptz  -- status update time
> ```
> 
> Do we wish to update this value every time we modify row(s) in the table or only when workflow status is updated?

That `status update time` comment may be copypasta because it is on every `updated` column in the schema
For example, in `ContinuousWorkloadInstance` it cannot be just for `status` because there is none in that table.
I think it represents the last time WFL "considered" that workload, and probably should be (ahem) updated
whenever anything in that tree of workload tables is changed -- including workflow status when that applies.

Reply:
That makes sense to me, thanks for confirming.  I think I've handled this now: changes in the data model meant that I need to update the source details instance with the snapshot refid, *and* the workload table with the updated time.